### PR TITLE
Switch bigint to bignumber.js

### PR DESCRIPTION
**Pull request by septs**

### CODE REVIEW

File: src/stories/RedPackets.tsx [hunk 1]
```diff
@@ -110,7 +111,7 @@ storiesOf('Plugin: Red Packets', module)
                         mnemonic: ['mem'],
                         name: 'Wallet Name',
                         passphrase: 'password',
-                        eth_balance: BigInt(2 * 10 ** 18),
+                        eth_balance: new BigNumber(2).multipliedBy(10).pow(18), // (2 * 10) ** 18
```
Comment (Jack-Works) [inline src/stories/RedPackets.tsx]: Wait... `2 * 10 ** 18` is `2 * (10 ** 18)`

**Comment by septs** [comment]
See #883

File: src/stories/RedPackets.tsx (hunk 1 above)
Comment (septs) [reply src/stories/RedPackets.tsx]: Thanks

File: src/plugins/Wallet/red-packet-fsm.ts [hunk 2]
```diff
@@ -83,12 +83,12 @@ export async function getRedPackets(owned?: boolean) {
 export async function createRedPacket(packet: createRedPacketInit): Promise<{ password: string }> {
     if (packet.send_total < packet.shares) {
         throw new Error('At least [number of red packets] tokens to your red packet.')
-    } else if (packet.shares < 0) {
+    } else if (packet.shares.isLessThan(0) /* packet.shares < 0n */) {
```
Comment (guanbinrui) [inline src/plugins/Wallet/red-packet-fsm.ts]: Does this lib requires compare between bns?

Comment (Jack-Works) [reply src/plugins/Wallet/red-packet-fsm.ts]: doesn't catch your meaning

Comment (guanbinrui) [reply src/plugins/Wallet/red-packet-fsm.ts]:
```javascript
packet.shares.isLessThan(new BigNumber(0))
```

Comment (septs) [reply src/plugins/Wallet/red-packet-fsm.ts]:
```typescript
packet.shares.isNegative()
```

fixed

File: src/components/InjectedComponents/PostDialog.tsx [hunk 3]
```diff
@@ -156,9 +157,10 @@ export function PostDialogUI(props: PostDialogUIProps) {
                                         props.onShareToEveryoneChanged(false)
                                     }
                                 }}
-                                label={`A Red Packet with $${formatBalance(BigInt(r.total), r.token?.decimals || 18)} ${
-                                    r.token?.name || 'ETH'
-                                } from ${r.sender.name}`}
+                                label={`A Red Packet with $${formatBalance(
+                                    new BigNumber(r.total),
+                                    r.token?.decimals ?? 18,
```
Comment (guanbinrui) [inline src/components/InjectedComponents/PostDialog.tsx]: `??` behavior different with `||`.

File: src/stories/RedPackets.tsx [hunk 4]
```diff
@@ -110,7 +111,7 @@ storiesOf('Plugin: Red Packets', module)
                         mnemonic: ['mem'],
                         name: 'Wallet Name',
                         passphrase: 'password',
-                        eth_balance: BigInt(2 * 10 ** 18),
+                        eth_balance: new BigNumber(2).multipliedBy(new BigNumber(10).pow(18)), // 2 * (10 ** 18)
```
Comment (guanbinrui) [inline src/stories/RedPackets.tsx]: Why not `2 * 10 ** 18`?

File: src/plugins/Wallet/UI/RedPacket/RedPacketDialog.tsx [hunk 5]
```diff
@@ -152,32 +150,25 @@ function NewPacketUI(props: RedPacketDialogProps & NewPacketProps) {
...
     const createRedPacket = () => {
+        const power = selectedTokenType.type === 'eth' ? 18 : selectedToken.decimals
         props.onCreateNewPacket({
-            duration: 60 /** seconds */ * 60 /** mins */ * 24 /** hours */,
+            duration: 60 /* seconds */ * 60 /* mins */ * 24 /* hours */,
             is_random: Boolean(is_random),
             network: rinkebyNetwork ? EthereumNetwork.Rinkeby : EthereumNetwork.Mainnet,
             send_message,
-            send_total: BigInt(send_total * 10 ** (selectedTokenType.type === 'eth' ? 18 : selectedToken!.decimals)),
+            send_total: new BigNumber(send_total).multipliedBy(new BigNumber(10).pow(power)),
```
Comment (guanbinrui) [inline src/plugins/Wallet/UI/RedPacket/RedPacketDialog.tsx]: What about `send_total * 10 ** power`?

File: src/stories/RedPackets.tsx (hunk 4 above)
Comment (septs) [reply src/stories/RedPackets.tsx]: see https://www.npmjs.com/package/bignumber.js#use 
```
new BigNumber(99999999999999999999)       // '100000000000000000000'
```

File: src/components/InjectedComponents/PostDialog.tsx (hunk 3 above)
Comment (septs) [reply src/components/InjectedComponents/PostDialog.tsx]: CC @Jack-Works

Comment (Jack-Works) [reply src/components/InjectedComponents/PostDialog.tsx]: Yes, it's intentional. The original `||` is wrong

File: src/stories/RedPackets.tsx (hunk 4 above)
Comment (Jack-Works) [reply src/stories/RedPackets.tsx]: (The original code is actually unsafe)

File: src/plugins/Wallet/wallet.ts [hunk 6]
```diff
@@ -18,21 +19,20 @@ export function getWalletProvider() {
...
-        return getWalletProvider()
-            .queryERC20TokenBalance(addr, erc20Addr)
-            .then((x) => onWalletERC20TokenBalanceUpdated(addr, erc20Addr, x))
-            .catch(() => {
-                // do nothing
-            })
+    async (addr: string, erc20Addr: string) => {
+        try {
+            const x = await getWalletProvider().queryERC20TokenBalance(addr, erc20Addr)
+            return onWalletERC20TokenBalanceUpdated(addr, erc20Addr, x)
```
Comment (Jack-Works) [inline src/plugins/Wallet/wallet.ts]: If you use `return`, error in `onWalletERC20TokenBalanceUpdated` won't be caught by the catch clause.

Comment (septs) [reply src/plugins/Wallet/wallet.ts]: ACK

File: src/plugins/Wallet/wallet.ts [hunk 7]
```diff
@@ -211,5 +211,16 @@ export async function onWalletERC20TokenBalanceUpdated(address: string, tokenAdd
 async function getWalletByAddress(t: IDBPSafeTransaction<WalletDB, ['Wallet'], 'readonly'>, address: string) {
     const rec = await t.objectStore('Wallet').get(address)
     assert(rec)
+    if (rec.eth_balance) {
+        Object.setPrototypeOf(rec.eth_balance, BigNumber.prototype)
```
Comment (Jack-Works) [inline src/plugins/Wallet/wallet.ts]: Store a string like "1000000000000000000000000000000000" into the database and read it like `new BigNumber(rec.eth_balance)`. If you choose to drop the prototype and store it directly, we will depend on the internal data structure of BigNumber forever.

File: src/plugins/Wallet/wallet.ts (hunk 6 above)
Comment (SunriseFox) [reply src/plugins/Wallet/wallet.ts]: It seems actually the same according to my JS knowledge...

Comment (Jack-Works) [reply src/plugins/Wallet/wallet.ts]: `return await expr` is ok

Comment (SunriseFox) [reply src/plugins/Wallet/wallet.ts]: Ah, so it is.

Comment (Jack-Works) [reply src/plugins/Wallet/wallet.ts]: @SunriseFox try the following code:
```js
async function a() {
    try { return Promise.reject() }
    catch { return 'ok' }
}
await a() // uncaught promise rejection
async function b() {
    try { return await Promise.reject() }
    catch { return 'ok' }
}
await b() //"ok"
```

File: src/plugins/Wallet/red-packet-fsm.ts [hunk 8]
```diff
@@ -1,21 +1,23 @@
+import _ from 'lodash'
```
Comment (Jack-Works) [inline src/plugins/Wallet/red-packet-fsm.ts]: hmm, tree shake, please...

Comment (Jack-Works) [reply src/plugins/Wallet/red-packet-fsm.ts]: And export from `lodash-es`

File: src/plugins/Wallet/wallet.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
+import _ from 'lodash'
```
Comment (Jack-Works) [inline src/plugins/Wallet/wallet.ts]: { ... } from lodash-es

Comment (septs) [reply src/plugins/Wallet/wallet.ts]: XD

File: src/plugins/Wallet/wallet.ts [hunk 10]
```diff
@@ -18,21 +20,16 @@ export function getWalletProvider() {
...
 const memoQueryERC20Token = memoizePromise(
-    (addr: string, erc20Addr: string) => {
-        return getWalletProvider()
-            .queryERC20TokenBalance(addr, erc20Addr)
-            .then((x) => onWalletERC20TokenBalanceUpdated(addr, erc20Addr, x))
-            .catch(() => {
-                // do nothing
-            })
+    async (addr: string, erc20Addr: string) => {
+        const x = await getWalletProvider().queryERC20TokenBalance(addr, erc20Addr)
```
Comment (Jack-Works) [inline src/plugins/Wallet/wallet.ts]: Actually the semantics has changed there. This await won't be caught by the noop, but in the original version it is caught by the `.catch` call.

File: src/plugins/Wallet/database/Wallet.db.ts [hunk 11]
```diff
@@ -9,9 +9,9 @@ export const createWalletDBAccess = createDBAccess(() => {
     return openDB<WalletDB>('maskbook-plugin-wallet', 1, {
         upgrade(db, oldVersion, newVersion, tx) {
             function v0_v1() {
-                db.createObjectStore('RedPacket', { keyPath: path<keyof RedPacketRecord>('id') })
-                db.createObjectStore('ERC20Token', { keyPath: path<keyof ERC20TokenRecord>('address') })
-                db.createObjectStore('Wallet', { keyPath: path<keyof WalletRecord>('address') })
+                db.createObjectStore('RedPacket', { keyPath: path<keyof RedPacketRecordInDatabase>('id') })
```
Comment (Jack-Works) [inline src/plugins/Wallet/database/Wallet.db.ts]: Does it be able to migrate old data (raw bigint) or support both of them?

Comment (septs) [reply src/plugins/Wallet/database/Wallet.db.ts]: need old data sample, verifying the issue.

**Pull request merged**